Sources/DeepLearning/Optimizer.swift

```diff
@@ -71,35 +76,29 @@ public class Adam<Model: Layer, Scalar: TensorFlowFloatingPoint>: Optimizer
         self.beta2 = beta2
         self.epsilon = epsilon
         self.decay = decay
-    }
 
-    public convenience init(
-        for _: __shared Model,
-        learningRate: Scalar = 1e-3,
-        beta1: Scalar = 0.9,
-        beta2: Scalar = 0.999,
-        epsilon: Scalar = 1e-8,
-        decay: Scalar = 0,
-        scalarType: Scalar.Type
-    ) {
-        self.init(
-            learningRate: learningRate,
-            beta1: beta1,
-            beta2: beta2,
-            epsilon: epsilon,
-            decay: decay)
+        // Initialize first & second moments to be zeros of the same shape.
+        // We can't use `Model.AllDifferentiableVariables.zero` due to the
+        // interaction between Key Paths and Differentiable Arrays.
+        firstMoments = model.allDifferentiableVariables
+        secondMoments = model.allDifferentiableVariables
+        for kp in firstMoments.recursivelyAllWritableKeyPaths(to: Tensor<Float>.self) {
+          firstMoments[keyPath: kp] = Tensor(zeros: firstMoments[keyPath: kp].shape)
```

Comment:
To reduce boilerplace, you can make a `fileprivate` extension method `formZeros()`.

```swift
fileprivate extension Tensor {
    mutating func formZeros() {
        self = Tensor(zeros: shape)
    }
}
```

Then you can just use 
```swift
 firstMoments[keyPath: kp].formZeros()
```

Reply:
Fixed (picked a slightly different name that I thought was clearer).